docs/firmware-architecture-and-design.md

```diff
@@ -0,0 +1,119 @@
+ *                                set the speed of this dog, returning the max 
+ *                                speed achieved, in m/s.
+ * @ereturn A pointer to a dog with the given parameters
+ */
+Dog_t* dog_create(DogColor color, float (*attempt_to_run_at_speed)(float)); 
+
+/**
+ * Destroy the given Dog, freeing any memory allocated for it
+ * @param dog The Dog to destroy
+ */
+void dog_destroy(Dog_t* dog);
+
+/**
+ * Get the color of a dog
+ * @param dog The dog to get the color of
+ * @return The color of the given dog
+ */
+DogColor dog_getColor(Dog_t* dog);
+
+/**
+ * Attempt to set the speed of a Dog
+ * @param desired_speed The speed to attempt, in m/s
+ * @return The maximum speed actually achieved, in m/s
+ */
+float dog_attemptToRunAtSpeed(Dog_t* dog, float desired_speed);
+```
+
+There are a few things to note here: 
+* We have declared a `struct Dog` and typedef'd it to just `Dog_t`, but we *haven't actually said what's in the struct*. This is a crucial point, because it means that anything that includes `dog.h` can only interact with a `struct Dog` through the functions we define here. 
+* We passed a function into `dog_create`. This is a really useful technique that is part of what enables us to do stuff like providing different "implementations" of a robot to the `app` layer depending on if we're in simulation or running on the real robots.
+
+We would implement this "class" as follows:
+
+_dog.c_
+``` C
+#include "dog.h"
+
+struct Dog {
+    DogColor color;
+    float (*attempt_to_run_at_speed)(float);
+};
+
+Dog_t* dog_create(DogColor color, float (*attempt_to_run_at_speed)(float)){
+    // NOTE: We can't do this outside this `.c` file, because the only place
+    //       that we know the size of the `Dog` struct is here.
+    Dog_t* dog = (Dog_t*)(malloc(sizeof(Dog_t)));
+
+    dog->color = color;
+    dog->attempt_to_run_at_speed = attempt_to_run_at_speed;
+
+    return dog;
+}
+
+void dog_destroy(Dog_t* dog){
+    // NOTE: We can't do this outside this `.c` file, because the only place
+    //       that we know the size of the `Dog` struct is here.
+    free(dog);
+}
+
+DogColor dog_getColor(Dog_t* dog){
+    return dog->color;
+}
+
+float dog_attemptToRunAtSpeed(Dog_t* dog, float desired_speed){
+    return dog->attempt_to_run_at_speed(desired_speed);
+}
+```
+
+and we would use this class as follows:
+``` C
+#include "dog.h"
+
+// Assume we get our "attempt_to_run_at_speed" method from this header
+#include "dog_simulator.h"
+
+Dog_t* my_dog = dog_create(BROWN, &simulated_dog_attempt_speed);
+
+assert(dog_getColor(my_dog) == BROWN);
+
+float achieved_speed = dog_attemptToRunAtSpeed(my_dog, 9001);
+
+dog_destroy(my_dog);
+```
+
+# App/IO Split
+The firmware is divided into `app` and `io` layers. 
+* The `app` layer contains anything that is portable between all our robots and our simulator. 
+* The `io` layer contains everything below this. 
```

Comment:
nit: perhaps an example of `io` would help? is `io` = `HAL?`

Reply:
Hmm, it's not `HAL`, it's everything above `HAL` and below `app`. Thoughts on the best way to say this?

Reply:
* The `io` layer contains everything below the `app` layer and above the `HAL` layer?
Can be more verbose since it's pretty short

Reply:
changed to 

* The `io` layer contains everything above this, down to and including calls to libraries like `HAL`.

Reply:
changed to 

* The `io` layer contains everything above this, down to and including calls to libraries like `HAL` (`HAL` is a library that provides thin wrappers around hardware).